x/photon/keeper/msg_server.go

```diff
@@ -0,0 +1,113 @@
+package keeper
+
+import (
+	"context"
+
+	"cosmossdk.io/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	govtypes "github.com/atomone-hub/atomone/x/gov/types"
+	"github.com/atomone-hub/atomone/x/photon/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+// MintPhoton implements the MsgServer.MintPhoton method.
+func (k msgServer) MintPhoton(goCtx context.Context, msg *types.MsgMintPhoton) (*types.MsgMintPhotonResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	params := k.GetParams(ctx)
+	if params.MintDisabled {
+		return nil, types.ErrMintDisabled
+	}
+
+	// Ensure burned amount denom is bond denom
+	bondDenom := k.stakingKeeper.BondDenom(ctx)
+	if msg.Amount.Denom != bondDenom {
+		return nil, types.ErrBurnInvalidDenom
+	}
+	// Compute photons to mint
+	var (
+		bondDenomSupply = k.bankKeeper.GetSupply(ctx, bondDenom).Amount.ToLegacyDec()
+		uphotonSupply   = k.bankKeeper.GetSupply(ctx, types.Denom).Amount.ToLegacyDec()
+		conversionRate  = k.conversionRate(ctx, bondDenomSupply, uphotonSupply)
+		bondDenomToBurn = msg.Amount
+		uphotonToMint   = bondDenomToBurn.Amount.ToLegacyDec().Mul(conversionRate).RoundInt()
+	)
+	// If no photon to mint, do not burn bondDenomToBurn, returns an error
+	if uphotonToMint.IsZero() {
+		return nil, types.ErrZeroMintPhotons
+	}
+	// If photonToMint + photonSupply > photonMaxSupply, returns an error
+	if uphotonSupply.Add(uphotonToMint.ToLegacyDec()).GT(sdk.NewDec(types.MaxSupply)) {
+		return nil, types.ErrNotEnoughPhotons
+	}
```

Comment:
Both checks but especially the second could be avoided. The second one especially, considering you get the conversionRate and its implicit  photonToMint + photonSupply <= photonMaxSupply because the conversionRate considers mintablePhotons = photonMaxSupply - photonMinted.

Seemengly, for the reasons explained in the ADR if the blockchain is producing blocks it means a non-zero amount of bond denom is staked, meaning it's impossible to convert **all** bond denom in cirulation and therefore the cap will never be reached.

Reply:
The first error can indeed happen due to rounding issue so it should stay, but the second check can be safely removed as agreed.

Reply:
I removed the second check [3c9199f](https://github.com/atomone-hub/atomone/pull/53/commits/3c9199fd5e9e9a473e5121cf02e6e21753b946d3).
As you noted the first one is required when the photon to mint becomes 0 after rounding.